test/unified-config.test.js

```diff
@@ -0,0 +1,453 @@
+  })
+
+  //
+  // restore the configuration
+  //
+  after (function () {
+    for (const k in savedEnv) {
+      process.env[k] = savedEnv[k];
+    }
+    try {fs.renameSync(`${rootConfigName}-renamed.json`, `${rootConfigName}.json`)} catch (e) {}
+    try {fs.renameSync(`${rootConfigName}-renamed.js`, `${rootConfigName}.js`)} catch (e) {}
+  })
+
+  //============================
+  // individual tests start here
+  //============================
+
+  it('should default correctly when no config file or env vars', function () {
+    const cfg = guc();
+
+    doChecks(cfg);
+  })
+
+  it('should set known config keys in a config file', function () {
+    const config = {
+      enabled: false,
+      hostnameAlias: 'bruce',
+      serviceKey: 'f'.repeat(64),
+      ignoreConflicts: true,
+      domainPrefix: 'fubar',
+    }
+    writeConfigJSON(config);
+
+    const cfg = guc();
+
+    doChecks(cfg, {global: config});
+  })
+
+  it('should warn about deprecated config file keys', function () {
+    const config = {
+      traceMode: 'always',
+      sampleRate: 1000000,
+    }
+    writeConfigJSON(config);
+
+    const cfg = guc();
+
+    // fix up to match translated value.
+    config.traceMode = 1;
+    const warnings = [
+      'traceMode is deprecated; it will be invalid in the future',
+      'sampleRate is deprecated; it will be invalid in the future',
+    ];
+    doChecks(cfg, {global: config, warnings});
+  })
+
+  it('should warn about unknown config keys', function () {
+    const config = {badKeyRising: 'i have been a bad key'};
+    writeConfigJSON(config);
+
+    const cfg = guc();
+
+    const unusedConfig = ['badKeyRising: i have been a bad key'];
+    doChecks(cfg, {unusedConfig});
+  })
+
+  it('should not set keys with invalid values', function () {
+    const config = {ec2MetadataTimeout: 'hello', createTraceIdsToken: 'bruce'};
+    writeConfigJSON(config);
+
+    const cfg = guc();
+
+    const warnings = [
+      'invalid configuration file value createTraceIdsToken: bruce',
+      'invalid configuration file value ec2MetadataTimeout: hello',
+    ];
+    doChecks(cfg, {warnings});
+  })
+
+  it('should handle a non-default config file correctly', function () {
+    const config = {ec2MetadataTimeout: 4000};
+    const filename = 'non-default.json';
+    fs.writeFileSync(filename, JSON.stringify(config));
+
+    process.env.APPOPTICS_APM_CONFIG_NODE = filename;
+
+    const cfg = guc();
+
+    fs.unlinkSync(filename);
+
+    doChecks(cfg, {file: `${process.cwd()}/${filename}`, global: config});
+  })
+
+  //
+  // config file errors
+  //
+  it('should correctly report an error reading a non-default config file', function () {
+    const file = process.env.APPOPTICS_APM_CONFIG_NODE = 'xyzzy-bruce.json';
+    const fullpath = `${process.cwd()}/${file}`;
+
+    const cfg = guc();
+
+    const message = 'Cannot find module \'/home/bruce/solarwinds/ao-agent/xyzzy-bruce.json\'';
```

Comment:
Should this be generalized for other filesystems/users? `${process.cwd()}/${file}`

Reply:
generalization makes sense - thanks. i already have constructed fullpath = `${process.cwd()}/${file}` so i'm just using that. now it will handle macs. i think windows accepts `/` even though they love to output `\` but i'll cross that bridge when i get to it.